azurerm/internal/services/apimanagement/api_management_gateway_api_resource.go

```diff
@@ -0,0 +1,162 @@
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceApiManagementGatewayApi() *pluginsdk.Resource {
+	return &pluginsdk.Resource{
+		Create: resourceApiManagementGatewayApiCreate,
+		Read:   resourceApiManagementGatewayApiRead,
+		Delete: resourceApiManagementGatewayApiDelete,
+
+		Importer: pluginsdk.ImporterValidatingResourceId(func(id string) error {
+			_, err := parse.GatewayApiID(id)
+			return err
+		}),
+
+		Timeouts: &pluginsdk.ResourceTimeout{
+			Create: pluginsdk.DefaultTimeout(30 * time.Minute),
+			Read:   pluginsdk.DefaultTimeout(5 * time.Minute),
+			Update: pluginsdk.DefaultTimeout(30 * time.Minute),
+			Delete: pluginsdk.DefaultTimeout(30 * time.Minute),
+		},
+
+		Schema: map[string]*pluginsdk.Schema{
+			"api_id": {
+				Type:         pluginsdk.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: azure.ValidateResourceID,
+			},
+			"gateway_id": {
+				Type:         pluginsdk.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: azure.ValidateResourceID,
+			},
+		},
+	}
+}
+
+func resourceApiManagementGatewayApiCreate(d *pluginsdk.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).ApiManagement.GatewayApisClient
+	ctx, cancel := timeouts.ForCreate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	apiID, err := parse.ApiID(d.Get("api_id").(string))
+	if err != nil {
+		return fmt.Errorf("parsing `api_id`: %v", err)
+	}
+
+	gatewayID, err := parse.GatewayID(d.Get("gateway_id").(string))
+	if err != nil {
+		return fmt.Errorf("parsing `gateway_id`: %v", err)
+	}
+
+	exists, err := client.GetEntityTag(ctx, gatewayID.ResourceGroup, gatewayID.ServiceName, gatewayID.Name, apiID.Name)
+	if err != nil {
+		if !utils.ResponseWasStatusCode(exists, http.StatusNoContent) {
+			if !utils.ResponseWasNotFound(exists) {
+				return fmt.Errorf("checking for presence of existing API %q / Gateway %q (API Management Service %q / Resource Group %q): %+v", apiID.Name, gatewayID, gatewayID.ServiceName, gatewayID.ResourceGroup, err)
+			}
+		}
+	}
+
+	if !utils.ResponseWasNotFound(exists) {
+		// TODO: can we pull this from somewhere?
+		subscriptionId := meta.(*clients.Client).Account.SubscriptionId
+		resourceId := fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.ApiManagement/service/%s/gateways/%s/apis/%s", subscriptionId, gatewayID.ResourceGroup, gatewayID.ServiceName, gatewayID, apiID.Name)
+		return tf.ImportAsExistsError("azurerm_api_management_gateway_api", resourceId)
+	}
+
+	params := &apimanagement.AssociationContract{}
+	resp, err := client.CreateOrUpdate(ctx, gatewayID.ResourceGroup, gatewayID.ServiceName, gatewayID.Name, apiID.Name, params)
+	if err != nil {
+		return fmt.Errorf("adding API %q to Gateway %q (API Management Service %q / Resource Group %q): %+v", apiID.Name, gatewayID.Name, gatewayID.ServiceName, gatewayID.ResourceGroup, err)
+	}
+
+	d.SetId(*resp.ID)
```

Comment:
We can use the subscription ID from the GatewayId struct, and also we'll want to use the ID constructor, and use the generated ID for the resource ID rather than the one returned by the API.

```suggestion
	id := parse.NewGatewayApiID(gatewayID.SubscriptionId, gatewayID.ResourceGroup, gatewayID.ServiceName, gatewayID.Name, apiID.Name)
	if !utils.ResponseWasNotFound(exists) {
		return tf.ImportAsExistsError("azurerm_api_management_gateway_api", id.ID())
	}

	params := &apimanagement.AssociationContract{}
	_, err = client.CreateOrUpdate(ctx, gatewayID.ResourceGroup, gatewayID.ServiceName, gatewayID.Name, apiID.Name, params)
	if err != nil {
		return fmt.Errorf("creating %s: %+v", id, err)
	}

	d.SetId(id.ID())
```